vlm/src/main/scala/geotrellis/contrib/vlm/MosaicRasterSource.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2018 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.contrib.vlm
+
+import cats.Semigroup
+import cats.implicits._
+import cats.data.NonEmptyList
+import geotrellis.vector._
+import geotrellis.raster._
+import geotrellis.raster.merge.Implicits._
+import geotrellis.raster.resample._
+import geotrellis.raster.reproject.Reproject
+import geotrellis.proj4.{CRS, WebMercator}
+import geotrellis.raster.io.geotiff.{AutoHigherResolution, OverviewStrategy}
+import geotrellis.raster.render._
+import geotrellis.spark.tiling.LayoutDefinition
+import geotrellis.util.GetComponent
+
+/**
+  * Single threaded instance of a reader for reading windows out of collections
+  * of rasters
+  *
+  * @param sources The underlying [[RasterSource]]s that you'll use for data access
+  * @param crs The crs to reproject all [[RasterSource]]s to anytime we need information about their data
+  * Since MosaicRasterSources represent collections of [[RasterSource]]s, we don't know in advance
+  * whether they'll have the same CRS. crs allows specifying the CRS on read instead of
+  * having to make sure at compile time that you're threading CRSes through everywhere correctly.
+  */
+case class MosaicRasterSource(sources: NonEmptyList[RasterSource], crs: CRS)
+    extends RasterSource {
+
+  import MosaicRasterSource._
+
+  /**
+    * Uri is required only for compatibility with RasterSource.
+    *
+    * It doesn't make sense to access "the" URI for a collection, so this throws an exception.
+    */
+  def uri: String = throw new NotImplementedError(
+    """
+      | MosaicRasterSources don't have a single uri. Perhaps you wanted the uri from
+      | one of this MosaicRasterSource's sources?
+    """.trim.stripMargin
+  )
+
+  val targetCellType = None
+
+  /**
+    * The bandCount of the first [[RasterSource]] in sources
+    *
+    * If this value is larger than the bandCount of later [[RasterSource]]s in sources,
+    * reads of all bands will fail. It is a client's responsibility to construct
+    * mosaics that can be read.
+    */
+  def bandCount: Int = sources.head.bandCount
+
+  def cellType: CellType = sources.head.cellType
+
+  def rasterExtent = {
+    val extents = sources map { _.reproject(crs).rasterExtent }
+    extents.tail.foldLeft(extents.head) { _ combine _ }
```

Comment:
[`RasterExtent.combine`](https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/RasterExtent.scala#L199) I'm not sure if this is expected behavior or not in this case. However the because underlying sources may not share the pixel alignment the result of `read` calls will be fudged in a surprising way where chain of `raster.reduce(raster)` is going to resample to potentially different pixel grid at each step.

Reply:
I think the fallout is that `MosaicRasterSource` should either be given a specific `RasterExtent` to which all input sources will be confirmed to or it should require that all sources already share pixel alignment.